File: leetcode/DP/leetcode_70_climbing_stairs/Solution.java
package DP.leetcode_70_climbing_stairs;

class Solution {
    public int climbStairs(int n) {
        if (n <= 1) {
            return 1;
        }

        int a = 1, b = 2;
        for (int i = 3; i < n+1; i ++) {
            int temp = a + b;
            a = b;
            b = temp;
            
        }
        return b;
        
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int result = solution.climbStairs(10);
        System.out.println(result);

    }
}

File: leetcode/DP/leetcode_70_climbing_stairs/solution.py
class Solution:
    def climbStairs(self, n: int) -> int:
        """
        n = 1, 1 method exists;
        n = 2, 2 methods exist;
        n = 3, 3 methods exist;
        n = 4, 5 methods exist;
        n = 5, 8 methods exist;
        """

        if n <= 1:
            return n
        
        """
        dp = [0] * (n+1)
        dp[1] = 1
        dp[2] = 2

        for i in range (3, n+1):
            dp[i] = dp[i-1] + dp[i-2]
        
        return dp[n]

        """
        a, b = 1, 2
        for i in range(3, n+1):
            a, b = b, a+b
        
        return b